### DISCUSSION TITLE
How to synchronize browser url whenredirecting on the server and route on the client?

### DISCUSSION THREAD

**Discussion by tobias-at-dm** [vercel/next.js#57329] [Help] [answered]
### Summary

I configured server-side redirection (via next.config.js), but during client-side routing (e.g. via Link), the browser url is not updated correctly.

How to I solve that properly?

### Additional information

```js
To solve this, I added a hook to the page (redirected to) to check for the browser location and replace the history state in case the location doesn't match the expected path. However, this seems like a bad approach.

So I'm wondering: Is there a way to do it properly?


useEffect(() => {
    if (window.location.pathname === "/origin-path") {
        window.history.replaceState({}, '', "/redirected-to-path");
    }
}, [])
```
```


### Example

_No response_

**Comment by HCG-LimSB** [comment]
You can use API Reference Functions [useRouter](https://nextjs.org/docs/pages/api-reference/functions/use-router)

```javascript
import { useRouter } from 'next/router'

const App = () => {
  const router = useRouter();

  useEffect(() => {
    if (router.route.pathname === '/origin-path') {
      router.push('/redirected-to-path');
    }
  }, [router.route]);

  return (
    <div>
      <h1>App</h1>
    </div>
  );
};

export default App;
```

**Reply by tobias-at-dm** [reply]
Thank you @HCG-LimSB 

But isn't that triggering a new client-side navigation?

**Reply by HCG-LimSB** [reply]
You're correct, using router.replace('/redirected-to-path') in the code I provided would trigger a new client-side navigation. If you want to update the browser's URL without causing a new client-side navigation, you can utilize JavaScript's window.history.replaceState method, similar to your initial approach. However, you can make this approach more robust and maintainable.

Here's a modified version of your code that uses window.history.replaceState without triggering a client-side navigation:

```javascript
useEffect(() => {
    if (window.location.pathname === '/origin-path') {
        const newUrl = '/redirected-to-path';
        window.history.replaceState({}, '', newUrl);
        // Also, update the current URL in the router object if necessary
        router.replace(newUrl, undefined, { shallow: true });
    }
}, []);
```
In this code, we first use window.history.replaceState to change the URL in the browser's history without triggering a new client-side navigation. Then, we update the current URL in the router object using router.replace. By setting { shallow: true }, we ensure that the URL is updated in the router without causing a full page reload or client-side navigation.

This approach maintains proper client-side routing while updating the URL as needed on the client side without starting a new navigation event.

**Reply by tobias-at-dm** [reply, marked as answer]
Thank you @HCG-LimSB (or ChatGPT ;) )! I'll use that for now. But then there is no more structural solution for this problem?

This/my solution doesn't scale very well. Plus, this is a part of the cross-cutting concern routing, which imo should be handled in a separate place, like a middleware or router configuration or something like that.
